src/utils/reporting_workflow.js

```diff
@@ -45,13 +50,130 @@ const IG_REGISTRY = {
+  const url = endpoint.address;
+  // Not sure how this will work since the url comes from endpoint.address....
+  const token = await connectToServer(url);
+  const headers = { Authorization: `Bearer ${token}` };
+  return axios
+    .post(url, bundle, { headers: headers })
+    .then(response => response.data)
+    .catch(err => console.error(err));
+}
+
+/**
+ * Helper function to call specific operations on the data trust service server.
+ *
+ * @param {string} operation = 'deidentify' | 'anonymize' | 'pseudonymize'
+ * @param {Bundle} bundle - the bundle to perform operation on
+ * @returns axios promise with data
+ */
+function dataTrustOperation(operation, bundle) {
+  return axios
+    .post(`${process.env.DATA_TRUST_SERVICE}/Bundle/$${operation}`, bundle)
+    .then(response => response.data)
+    .catch(err => console.error(err));
+}
+
+/**
+ * Function to use the PlanDefinition and triggering resource to create a report
+ *
+ * @param {PlanDefinition} planDef - the PlanDefinition resource
+ * @param {*} resource - the resource which triggered the notification
+ */
+async function initializeReportingWorkflow(planDef, resource = null) {
+  // TODO: MEDMORPH-49 to make sure the resource is always included
+  if (!resource) return;
+
+  let patient = null;
+  if (resource.patient) patient = await getReferencedResource(resource.patient.reference);
+  else if (resource.subject) patient = await getReferencedResource(resource.subject.reference);
+  else if (resource.resourceType === 'Patient') patient = resource;
+
+  let encounter = null;
+  if (resource.encounter) encounter = await getReferencedResource(resource.encounter.reference);
+  else if (resource.context) encounter = await getReferencedResource(resource.context.reference);
+  else if (resource.resourceType === 'Encounter') encounter = resource;
+
+  // QUESTION: Should encounter and patient be saved to the database?
+
+  const context = initializeContext(planDef, patient, encounter);
+  executeWorkflow(context);
+}
+
+/**
+ * Create the initial workflow context
+ *
+ * @param {PlanDefinition} planDefinition - the PlanDefinition resource for this workflow
+ * @param {Patient} patient - the patient resource from the triggering resource, null if not found
+ * @param {Encounter} encounter - the encounter resource from the trigger resource,
+ *    null if not found
+ * @returns the initialized context object
+ */
+function initializeContext(planDefinition, patient, encounter) {
+  const records = [];
+  if (patient) records.push(patient);
+  if (encounter) records.push(encounter);
+  const actionSequence = determineActionSequence(planDefinition);
+  const initialAction = planDefinition.action.find(a => a.id === actionSequence[0]);
+
   return {
     id: uuidv4(),
+    patient: patient,
+    records: records,
+    encounter: encounter,
+    planDefinition,
+    action: initialAction,
+    messageHeader: null, // Note: not used in actions
+    client: {
```

Comment:
Does the reference SMART on FHIR client support the things we need to do with backend service authorization / custom operations? https://github.com/smart-on-fhir/client-js
If it does then I'd propose we use that rather than invent new things here.

Reply:
After looking through the docs and a little bit of the code it _might_, but I only see stuff regarding the typical SMART launch not the backend auth. 
Either way I don't really think its necessary since we're not really "inventing" anything. The only extra steps are:
1. Get access token
2. Use access token

The FHIR client definitely supports 2 but adding the authorization header to a a FHIR client call is no easier than attaching it as a header to the axios call.

As for getting the access token, even if the FHIR client did support this for backend auth we already have all of the code implemented to get the access token and Mike is working on caching it. While this isn't necessarily the best defense, we have it and it works so I'm personally hesitant to switch.